[Compute] Add Managed Identity Support in Azure Disk Encryption for VMSS

anshuljain26 opened the pull request:
**Related command**

**Description**<!--Mandatory-->
Azure Disk Encryption (ADE) is adding support for using managed identity to authenticate to customer's keyvault.
As part of it, a new field (EncryptionIdentity) has been added to the VMSS model. By setting this field customer will be notifying ADE to use that managed identity for keyvault operations. The identity should also be explicitly assigned to the VMSS.

This PR adds a new parameter (EncryptionIdentity) to az vmss encryption enable cmdlet. If the parameter is present then the cmdlet will update the EncryptionIdentity field.

Encryption Identity field is also updated during VMSS creation if the encryption identity is a part of the identities assigned to the vmss

**Testing Guide**

**History Notes**
[Compute] `az vmss create`: Add `--encryption-identity` parameter to use that managed identity for Azure disk encryption
[Compute] `az vmss encryption enable`: Add `--encryption-identity` parameter to update or set encryption identity for Azure disk encryption

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

anshuljain26 commented:
@vimish Please review the PR

yonzhan commented:
Compute

evelyn-ys commented:
/azp run

zhoxing-ms commented:
Could you please refer to this document https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md to design commands and parameters related to managed identity?

anshuljain26 commented:
> Could you please refer to this document https://github.com/Azure/azure-cli/blob/dev/doc/managed_identity_command_guideline.md to design commands and parameters related to managed identity?

We have added a new Parameter encryption identity to authenticate to customer's keyvault.
As part of it, a new field (EncryptionIdentity) has been added to the VM model. By setting this field customer will be notifying ADE to use that managed identity for keyvault operations. And we want to set this field during vm creation or update it during the encryption. We have set this field because vm can have multiple identity assigned, so which identity we need to use for keyVault operation, we have added this new field

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/vm/_help.py:
```diff
@@ -2643,6 +2643,9 @@
         az vmss create -n MyVmss -g rg1 --image Debian11 --assign-identity  /subscriptions/99999999-1bf0-4dda-aec3-cb9272f09590/resourcegroups/myRG/providers/Microsoft.ManagedIdentity/userAssignedIdentities/myID
+  - name: Create a vmss with user assigned identity and add encryption identity for Azure disk encryption
+    text: >
+        az vmss create -n MyVm -g rg1 --image Debian11 --assign-identity myID --encryption-identity /subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/myRG/providers/Microsoft.ManagedIdentity/userAssignedIdentities/myID --orchestration-mode Uniform --lb-sku Standard
```
May I ask what is the difference between the `--assign-identity` and `--encryption-identity`?

anshuljain26 replied:
--assign-identity params is used to assign system or user identities associated with the Virtual Machine Scale set

--encryption-identity params is used to set which Identity used by ADE to get access token for keyvault operations.

zhoxing-ms replied:
Got it, thanks~

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/vm/custom.py:
```diff
@@ -3532,6 +3532,35 @@ def _get_public_ip_address_allocation(value, sku):
                 master_template.add_resource(build_msi_role_assignment(vmss_name, vmss_id, identity_role_id,
                                                                        role_assignment_guid, identity_scope, False))
+        if encryption_identity:
+            if not cmd.supported_api_version(min_api='2023-09-01', resource_type=ResourceType.MGMT_COMPUTE):
+                raise CLIError("Usage error: Encryption Identity required API version 2023-09-01 or higher."
+                               "You can set the cloud's profile to use the required API Version with:"
+                               "az cloud set --profile latest --name <cloud name>")
```
please set the `min_api='2023-09-01'` in the parameter definition instead of adding checking here

anshuljain26 replied:
Addressed Suggestion

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/vm/custom.py:
```diff
@@ -3532,6 +3532,35 @@ def _get_public_ip_address_allocation(value, sku):
+                    vmss_resource['properties']['virtualMachineProfile']['securityProfile']['encryptionIdentity'] \
+                        = vmss_securityProfile_EncryptionIdentity
+            else:
+                raise CLIError("Encryption Identity should be an ARM Resource ID of one of the "
+                               "user assigned identities associated to the resource")
```
Please use the specific error type `ArgumentUsageError` in stead of `CLIError`
```suggestion
                raise ArgumentUsageError("Encryption Identity should be an ARM Resource ID of one of the "
                               "user assigned identities associated to the resource")
```

anshuljain26 replied:
During VMSS creation if there is any exception, it will throw cliError exception by default, that's why used that cliError exception here

zhoxing-ms replied:
In fact, we usually recommend using more specific error type if they can be clearly classified, as this will help us in future Telemetry data analysis

zhoxing-ms replied:
In fact, we usually recommend using more specific error type if they can be clearly classified, as this will help us in future Telemetry data analysis

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/vm/_params.py:
```diff
@@ -1191,8 +1190,13 @@ def load_arguments(self, _):
-        c.argument('encryption_identity', help='Resource Id of the user managed identity which can be used for Azure disk encryption')
+    for scope in ['vm create', 'vmss create']:
+        with self.argument_context(scope) as c:
+            c.argument('encryption_identity', help='Resource Id of the user managed identity which can be used for Azure disk encryption')
```
```suggestion
            c.argument('encryption_identity', help='Resource Id of the user managed identity which can be used for Azure disk encryption', min_api='2023-09-01')
```

anshuljain26 replied:
Addressed Suggestion

zhoxing-ms replied:
Don't we need to add `min_api` for `vm create` and `vm encryption enable`?

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/vm/_params.py:
```diff
@@ -1191,8 +1190,13 @@ def load_arguments(self, _):
+
+    for scope in ['vm encryption enable', 'vmss encryption enable']:
+        with self.argument_context(scope) as c:
+            c.argument('encryption_identity', help='Resource Id of the user managed identity which can be used for Azure disk encryption')
```
```suggestion
            c.argument('encryption_identity', help='Resource Id of the user managed identity which can be used for Azure disk encryption', min_api='2023-09-01')
```

anshuljain26 replied:
Addressed Suggestion

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/vm/_validators.py:
```diff
@@ -1374,6 +1374,14 @@ def _validate_trusted_launch(namespace):
         namespace.enable_secure_boot = True
 
 
+def _validate_encryption_identity(cmd):
+    from azure.cli.core.profiles import ResourceType
+    if not cmd.supported_api_version(min_api='2023-09-01', resource_type=ResourceType.MGMT_COMPUTE):
+        raise CLIError("Usage error: Encryption Identity required API version 2023-09-01 or higher."
+                       "You can set the cloud's profile to use the required API Version with:"
+                       "az cloud set --profile latest --name <cloud name>")
```
Please use the `min_api` flag in parameter definition instead of adding the validation logic

anshuljain26 replied:
Removed the logic here and added min_api in parameter definition

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/vm/disk_encryption.py:
```diff
@@ -459,6 +496,13 @@ def encrypt_vmss(cmd, resource_group_name, vmss_name,  # pylint: disable=too-man
     if key_encryption_key:
         key_encryption_keyvault = key_encryption_keyvault or disk_encryption_keyvault
 
+    if encryption_identity and isVersionSuppprtedForEncryptionIdentity(cmd):
```
> isVersionSuppprtedForEncryptionIdentity

May I ask do we still need this method?

anshuljain26 replied:
Okay, will remove this condition checker from here

yanzhudd commented on src/azure-cli/azure/cli/command_modules/vm/_params.py:
```diff
@@ -1191,8 +1190,13 @@ def load_arguments(self, _):
 
-    with self.argument_context('vm encryption enable') as c:
-        c.argument('encryption_identity', help='Resource Id of the user managed identity which can be used for Azure disk encryption')
+    for scope in ['vm create', 'vm encryption enable']:
+        with self.argument_context(scope) as c:
+            c.argument('encryption_identity', help='Resource Id of the user managed identity which can be used for Azure disk encryption')
```
May I ask if the `min_api='2023-09-01'` restriction of `encryption_identity` does not applicable to the `vm create` and `vm encryption enable` commands?

anshuljain26 replied:
Addressed the suggestion
